NEW Add an extension to dynamically generate edit URLs

GuySartorelli opened the pull request:
Currently getting an edit link for an object is a bit of a pain - you have to generate a new link based on the predictable gridfield or other admin edit paths. This is partly improved by https://github.com/silverstripe/silverstripe-admin/pull/1360 - but that still doesn't resolve the problem for nested setups (e.g. object in a page's gridfield).

## NOTE
CI won't go green until https://github.com/silverstripe/silverstripe-admin/pull/1360 is merged in as the new tests rely on `ModelAdmin` having a `getEditLinkForManagedDataObject()` method

## Related issue
- https://github.com/silverstripe/silverstripe-admin/issues/1354

GuySartorelli committed:
NEW Add an extension to dynamically generate edit URLs

michalkleiner commented on code/HasCanonicalEditLink.php:
```diff
@@ -0,0 +1,152 @@
+
+            if ($remoteField) {
+                if ($remoteField === $reciprocalRelation) {
+                    if ($reciprocalRelation === 'reciprocalRelation') {
+                        var_dump([
+                            'HERE',
+                            'relation' => $relation
+                        ]);
+                    }
```
Should this be removed?

ChloeSartorelli replied:
Yes, oops. Removed now

michalkleiner commented on code/HasCanonicalEditLink.php:
```diff
@@ -0,0 +1,152 @@
+    /**
+     * Get a link to edit this page in the CMS.
+     */
+    public function CMSEditLink(): string
```
`File` already has `CMSEditLink` method, would this be a problem when applied to e.g. Image or other class extending `File`?

ChloeSartorelli replied:
I can't think of literally any reason someone would do that - but if they did, I'm pretty sure the trait implementation would take precedence.

michalkleiner commented on code/HasCanonicalEditLink.php:
```diff
@@ -0,0 +1,152 @@
+ * If the canonical edit owner is a has_one relation, the class on the other end
+ * of the relation must have a CMSEditLink() method.
+ */
+trait HasCanonicalEditLink
```
I was out for a bit and may have missed it, but do we have a naming convention for traits?

I'm a proponent of suffixing the name with the type of class/construct it is, e.g. `SomethingTrait`, `PageController`, `FieldExtension` etc. so it's immediately obvious what it is.

```suggestion
trait CanonicalEditLinkTrait
```

ChloeSartorelli replied:
We do have quite a few traits already - some are suffixed with `Trait` but the majority are not.
I think the fact that you declare it _as_ a trait, and you `use` it, make it abundantly clear it's a trait.

`Extension` and `Controller` are a bit different because those aren't built-in language structure things like a trait (we don't say `SiteTreeClass` after all).

emteknetnz replied:
Agree we shouldn't suffix `Trait` on this one (even though I'd actually like it, I also like the `Interface` suffix), because most of the code base doesn't suffix it

I don't like the use of the word `Canonical` at all - it really should just be `CMSEditLink`

I had a look at different traits in the codebase to find a common naming convention, there isn't that much in the way of consistency.

The most common pattern I could find was just calling the trait the feature it implemented e.g.
- trait ImageManiputlation
- trait FlushScheduler
- trait FormMessage

The next most consistent, though less common one was the Configurable / Extensible / Injectable - though those are less "concrete" concepts than just the other traits that just bolt a feature on

This makes a couple of obvious candidates
- trait CMSEditLink
- trait CMSEditLinkable (weird IMO)

I think `CMSEditLink` is the most non-horrible of the two

```php
class MyClass
{
   use CMSEditLink;
```
And I get a `CMSEditLink()` method

ChloeSartorelli replied:
I'll rename the trait to `CMSEditLink`

michalkleiner commented:
I remember having issues not being able to get an edit link to objects in past so this would be a welcome addition, however I'm not sure if it should be a trait or somehow integrated more closely with `DataObject` itself or via an automatically applied extension to `DataObject` etc. What is the benefit of using the trait over an extension? Is this opt-in or would it be automatically applied?

GuySartorelli commented:
> What is the benefit of using the trait over an extension?

Less (if any) overhead is the main benefit - but mostly it just didn't really occur to me to make this as an extension.

> Is this opt-in or would it be automatically applied?

It _has_ to be opt-in. There's no way to know the canonical parent of some arbitrary model. How would we differentiate between the 12 has_one relations and the 3 model admins that all have edit links for some record?
These numbers are arbitrary but that's kinda my point. There's no way for us to determine, at a product-level, what some arbitrary project's model's canonical edit context is.

GuySartorelli commented:
I've renamed the trait to `CMSEditLink` and renamed the `getEditLinkForObject()` method to `getEditLinkForDataObject()` to match the change in https://github.com/silverstripe/silverstripe-admin/pull/1360

GuySartorelli commented:
Steve and I have talked about this offline - I'll be converting this into an `Extension` and renaming `getEditLinkForDataObject()` to `getEditLinkForManagedDataObject()` so it's more intuitive.

GuySartorelli committed:
FIX Always add dataClass to models spec

ChloeSartorelli commented on code/ModelAdmin.php:
```diff
@@ -538,9 +538,14 @@ public function getManagedModels()
                 unset($models[$k]);
+            // Custom tab url segment but no custom title
+            } elseif (is_a($v, DataObject::class, true)) {
+                $models[$k] = ['dataClass' => $v, 'title' => singleton($v)->i18n_plural_name()];
```
This fix was needed for tests to go green.

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+            && $field->getConfig()->getComponentByType(GridFieldDetailForm::class);
+    }
+
+    private function constructLink(string $relation, int $id): string
```
```suggestion
    private function constructCMSEditLink(string $relation, int $id): string
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+        return Director::absoluteURL($relativeLink);
+    }
+
+    private function getLinkForRelation(array $componentConfig, DataObject $obj, string $reciprocalRelation, FieldList $fields): string
```
```suggestion
    private function getCMSEditLinkForRelation(array $componentConfig, DataObject $obj, string $reciprocalRelation, FieldList $fields): string
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+     * @throws LogicException if a link cannot be established
+     * e.g. if the object is not in a has_many relation or not edited inside a GridField.
+     */
+    public function getEditLinkForManagedDataObject(DataObject $obj, string $reciprocalRelation): string
```
```suggestion
    public function getCMSEditLinkForManagedDataObject(DataObject $obj, string $reciprocalRelation): string
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+ */
+class CMSEditLinkExtension extends Extension
+{
+    private static string $canonical_edit_owner = '';
```
```suggestion
    private static string $cms_edit_owner = '';
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+     *
+     * @return LeftAndMain|DataObject|null
+     */
+    public function getCanonicalEditOwner()
```
```suggestion
    public function getCMSEditOwner()
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+     */
+    public function getCanonicalEditOwner()
+    {
+        $ownerType = $this->owner->config()->get('canonical_edit_owner');
```
```suggestion
        $ownerType = $this->owner->config()->get('cms_edit_owner');
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+        }
+
+        if ($owner instanceof DataObject) {
+            $relativeLink = $owner->getEditLinkForManagedDataObject($this->owner, $this->owner->config()->get('canonical_edit_owner'));
```
```suggestion
            $relativeLink = $owner->getCMSEditLinkForManagedDataObject($this->owner, $this->owner->config()->get('canonical_edit_owner'));
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+     */
+    public function CMSEditLink(): string
+    {
+        $owner = $this->owner->getCanonicalEditOwner();
```
```suggestion
        $owner = $this->owner->getCMSEditOwner();
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+            return '';
+        }
+
+        if (!$owner->hasMethod('getEditLinkForManagedDataObject')) {
```
```suggestion
        if (!$owner->hasMethod('getCMSEditLinkForManagedDataObject')) {
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+        }
+
+        if (!$owner->hasMethod('getEditLinkForManagedDataObject')) {
+            throw new LogicException('The canonical owner must implement getEditLinkForManagedDataObject()');
```
```suggestion
            throw new LogicException('The canonical owner must implement getCMSEditLinkForManagedDataObject()');
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+        if ($owner instanceof DataObject) {
+            $relativeLink = $owner->getEditLinkForManagedDataObject($this->owner, $this->owner->config()->get('canonical_edit_owner'));
+        } else {
+            $relativeLink = $owner->getEditLinkForManagedDataObject($this->owner);
```
```suggestion
            $relativeLink = $owner->getCMSEditLinkForManagedDataObject($this->owner);
```

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+
+    private function constructLink(string $relation, int $id): string
+    {
+        $ownerType = $this->owner->config()->get('canonical_edit_owner');
```
```suggestion
        $ownerType = $this->owner->config()->get('cms_edit_owner');
```

emteknetnz reviewed (changes requested):
Mostly naming stuff

Key thing that we missed when we merged https://github.com/silverstripe/silverstripe-admin/pull/1360 was the method name ModelAdmin::getEditLinkForManagedDataObject - it should be getCMSEditLinkForManagedDataObject for consistency with the rest of the code base

A quick string search

vendor/silverstripe/**/*.php

"CMSEditLink(" - 33 results
"EditLink(" - 37 results

Of the "EditLink(" that weren't "CMSEditLink(":
"updateCMSEditLink(" - 1 result
"getFileEditLink(" - 3 results

So you'll also need to rename https://github.com/silverstripe/silverstripe-admin/blob/1/code/ModelAdmin.php#L230 in this PR as well

The other thing is the word "Canonical" is very confusing.  Rename this to "CMSEdit" so that it's standard with everything else

I've left a bunch of recommended edits, though there's a large number of things that will need renaming e.g. the test files, docs, etc

GuySartorelli commented:
> The other thing is the word "Canonical" is very confusing. Rename this to "CMSEdit" so that it's standard with everything else

"canonical" _is_ correct in this context - e.g. the `<link rel="canonical" href="https://www.example.com/page/" />` meta tag tells search engines what the _canonical_ URL is for a page, in case the page can be accessed via multiple URLs.

But I'll make that change to avoid peer review ping pong and to match the other changes where "canonical" was already changed.

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    /**
+     * Get a link to edit this page in the CMS.
```
```suggestion
     * Get a link to edit this DataObject in the CMS.
```

ChloeSartorelli replied:
Updated

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+ *
+ * Note that the cms edit owner must implement a getCMSEditLinkForManagedDataObject() method.
+ *
+ * If the cms edit owner is a has_one relation, the class on the other end
```
```suggestion
 * If the cms_edit_owner is a has_one relation, the class on the other end
```

ChloeSartorelli replied:
done

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+    private static string $cms_edit_owner = '';
+
+    /**
+     * Get the admin or DataObject which owns this object for CMS editing purposes.
```
```suggestion
     * Get the ModelAdmin, LeftAndMain or DataObject which owns this object for CMS editing purposes.
```

ChloeSartorelli replied:
ModelAdmin _is_ a LeftAndMain - and they're all admin. I'll make this change but it's super subjective.

ChloeSartorelli replied:
done

emteknetnz commented on code/CMSEditLinkExtension.php:
```diff
@@ -0,0 +1,149 @@
+    /**
+     * Get the admin or DataObject which owns this object for CMS editing purposes.
+     *
+     * @return LeftAndMain|DataObject|null
```
```suggestion
     * @return LeftAndMain|ModelAdmin|DataObject|null
```

ChloeSartorelli replied:
ModelAdmin is a LeftAndMain so that's not necessary here. I'll make the change for the text comment but it's not appropriate for a type hint.

emteknetnz commented on tests/php/CMSEditLinkExtensionTest/BelongsToModelAdmin.php:
```diff
@@ -0,0 +1,37 @@
+use SilverStripe\Dev\TestOnly;
+use SilverStripe\ORM\DataObject;
+
+class BelongsToModelAdmin extends DataObject implements TestOnly
```
Could you please rename this, it's not a "belongs_to" relationship, which is defined as the reciprocal of a has_one relationship when it's a 1-1 relations.

Also, it being suffixed with "ModelAdmin" confused me for a long time because I mistakenly thought it extended `ModelAdmin`

Even just calling this `TestDataObject` would be an improvement

ChloeSartorelli replied:
That wouldn't be an improvement, it's a lot less descriptive. I'll rename it to something that doesn't contain "belongsto" or end with "modeladmin". But given this is a test object I don't see that it matters _all_ that much what it's called.

ChloeSartorelli replied:
Done

emteknetnz commented on tests/php/CMSEditLinkExtensionTest/BelongsToModelAdmin.php:
```diff
@@ -0,0 +1,37 @@
+        'Nested' => NestedObject::class . '.Parent',
+        'AnotherArbitraryRelation' => NestedObject::class . '.AnotherOfTheSameClass',
+        'BasicNested' => BasicNestedObject::class,
+        'PolyMorphic' => PolymorphicNestedObject::class,
```
```suggestion
        'Polymorphic' => PolymorphicNestedObject::class,
```

ChloeSartorelli replied:
Done
